File: src/vinfo/providers/docker_provider.py
import asyncio
from vinfo.constants import HealthStatus
from vinfo.core.base import BaseEnvironmentProvider
from vinfo.core.models import DockerInfo, EnvironmentInfo
from vinfo.core.registry import register
from vinfo.core.runner import CommandRunner


@register
class DockerProvider(BaseEnvironmentProvider):
    name = "docker"
    display_name = "Docker"
    icon = "\U0001f30a"
    priority = 40

    def __init__(self, runner: CommandRunner):
        self._runner = runner

    async def detect(self) -> bool:
        return self._runner.which("docker") is not None

    async def collect(self) -> DockerInfo:
        docker_bin = self._runner.which("docker") or ""
        docker_ver = await self._runner.run(f"{docker_bin} --version 2>&1")
        version = docker_ver.stdout.replace("Docker version ", "").split(",")[0].strip() if docker_ver.success else "Unknown"

        info = DockerInfo(
            name=self.name,
            display_name=f"Docker {version}",
            status=HealthStatus.UNKNOWN,
            version=version,
            path=docker_bin,
        )

        daemon_check = await self._runner.run(
            f"{docker_bin} info --format '{{{{.ServerVersion}}}}' 2>&1", timeout=5
        )
        if daemon_check.success:
            info.daemon_running = True
            info.status = HealthStatus.HEALTHY
        else:
            info.daemon_running = False
            info.status = HealthStatus.WARNING
            info.errors.append("Docker daemon is not running")

        if info.daemon_running:
            containers_res, images_res, volumes_res = await asyncio.gather(
                self._runner.run(f"{docker_bin} ps -a --format '{{{{.ID}}}}\\t{{{{.Names}}}}\\t{{{{.Image}}}}\\t{{{{.Status}}}}' 2>&1"),
                self._runner.run(f"{docker_bin} images --format '{{{{.ID}}}}\\t{{{{.Repository}}}}\\t{{{{.Tag}}}}\\t{{{{.Size}}}}' 2>&1"),
                self._runner.run(f"{docker_bin} volume ls --format '{{{{.Name}}}}\\t{{{{.Driver}}}}' 2>&1"),
            )

            if containers_res.success:
                info.containers = []
                for line in containers_res.stdout.splitlines():
                    parts = line.split("\t")
                    if len(parts) >= 4:
                        info.containers.append({
                            "id": parts[0],
                            "name": parts[1],
                            "image": parts[2],
                            "status": parts[3],
                        })

            if images_res.success:
                info.images = []
                for line in images_res.stdout.splitlines():
                    parts = line.split("\t")
                    if len(parts) >= 4:
                        info.images.append({
                            "id": parts[0],
                            "repository": parts[1],
                            "tag": parts[2],
                            "size": parts[3],
                        })

            if volumes_res.success:
                info.volumes = []
                for line in volumes_res.stdout.splitlines():
                    parts = line.split("\t")
                    if len(parts) >= 2:
                        info.volumes.append({"name": parts[0], "driver": parts[1]})

        compose_res, context_res = await asyncio.gather(
            self._runner.run(f"{docker_bin} compose version 2>&1"),
            self._runner.run(f"{docker_bin} context show 2>&1"),
        )
        if compose_res.success:
            first_line = compose_res.stdout.splitlines()[0] if compose_res.stdout else ""
            info.compose_version = first_line.strip() or None
        if context_res.success:
            info.context_name = context_res.stdout.strip()

        return info

    async def health_check(self) -> tuple[HealthStatus, list[str]]:
        if not self._runner.which("docker"):
            return HealthStatus.NOT_FOUND, ["Docker is not installed"]
        return HealthStatus.HEALTHY, []

    def get_quick_actions(self) -> list[dict]:
        return [
            {"id": "prune_images", "label": "Prune dangling images", "description": "Remove unused dangling images", "dangerous": True},
            {"id": "prune_volumes", "label": "Prune unused volumes", "description": "Remove unused volumes", "dangerous": True},
            {"id": "list_containers", "label": "List all containers", "description": "Show all containers (running + stopped)", "dangerous": False},
        ]

File: VinfoBar/Sources/VinfoBar/Providers/DockerProvider.swift
import Foundation

final class DockerProvider: EnvironmentProvider {
    static let name = "docker"
    static let displayName = "Docker"
    static let icon = "docker"
    static let priority = 40

    private let runner = CommandRunner.shared

    func detect() async -> Bool {
        return runner.which("docker") != nil
    }

    func collect() async -> any EnvironmentInfo {
        // Version info
        let verResult = await runner.run("docker --version")

        var version = "Unknown"
        if verResult.success {
            // Extract version from "Docker version 24.0.5, build ..."
            let pattern = "Docker version ([0-9.]+)"
            if let range = verResult.stdout.range(of: pattern, options: .regularExpression) {
                version = verResult.stdout[range].replacingOccurrences(of: "Docker version ", with: "")
            }
        }

        var info = DockerInfo(displayName: "Docker \(version)")
        info.version = version
        info.path = runner.which("docker")

        // Check daemon (short timeout)
        let daemonResult = await runner.run("docker info --format '{{.ServerVersion}}' 2>/dev/null", timeout: 5)
        info.daemonRunning = daemonResult.success

        if info.daemonRunning {
            // Get containers, images, volumes in parallel
            async let containersResult = runner.run("docker ps -a --format '{{.ID}}\t{{.Names}}\t{{.Image}}\t{{.Status}}'")
            async let imagesResult = runner.run("docker images --format '{{.ID}}\t{{.Repository}}\t{{.Tag}}\t{{.Size}}'")
            async let volumesResult = runner.run("docker volume ls --format '{{.Name}}\t{{.Driver}}'")
            async let composeResult = runner.run("docker compose version --short 2>/dev/null || true")
            async let contextResult = runner.run("docker context show")

            let containers = await containersResult
            let images = await imagesResult
            let volumes = await volumesResult
            let compose = await composeResult
            let context = await contextResult

            // Parse containers
            if containers.success {
                info.containers = containers.stdout.split(separator: "\n").compactMap { line -> DockerContainer? in
                    let parts = line.split(separator: "\t", maxSplits: 3)
                    if parts.count >= 4 {
                        return DockerContainer(
                            containerId: String(parts[0]),
                            name: String(parts[1]),
                            image: String(parts[2]),
                            status: String(parts[3])
                        )
                    }
                    return nil
                }
            }

            // Parse images
            if images.success {
                info.images = images.stdout.split(separator: "\n").compactMap { line -> DockerImage? in
                    let parts = line.split(separator: "\t", maxSplits: 3)
                    if parts.count >= 4 {
                        return DockerImage(
                            imageId: String(parts[0]),
                            repository: String(parts[1]),
                            tag: String(parts[2]),
                            size: String(parts[3])
                        )
                    }
                    return nil
                }
            }

            // Parse volumes
            if volumes.success {
                info.volumes = volumes.stdout.split(separator: "\n").compactMap { line -> DockerVolume? in
                    let parts = line.split(separator: "\t")
                    if parts.count >= 2 {
                        return DockerVolume(name: String(parts[0]), driver: String(parts[1]))
                    }
                    return nil
                }
            }

            if compose.success && !compose.stdout.isEmpty {
                info.composeVersion = compose.stdout
            }

            if context.success {
                info.contextName = context.stdout
            }

            info.status = .healthy
        } else {
            info.status = .warning
            info.errors.append("Docker daemon not running")
        }

        return info
    }

    func healthCheck() async -> (HealthStatus, [String]) {
        if runner.which("docker") == nil {
            return (.notFound, ["Docker not installed"])
        }
        return (.healthy, [])
    }

    func quickActions() -> [ActionDefinition] {
        [
            ActionDefinition(id: "list_containers", label: "List Containers", description: "Show all containers", isDangerous: false),
            ActionDefinition(id: "list_images", label: "List Images", description: "Show all images", isDangerous: false),
            ActionDefinition(id: "prune_images", label: "Prune Images", description: "Remove dangling images", isDangerous: true),
        ]
    }
}